### PR TITLE
Refactor: Configuration

### PR DESCRIPTION
**Pull request by typerandom**
Refactors the way that configuration is handled so that configuration that was previously in the SDK is now in the integration.

**Important:** This may only be merged once there is a new release for [the stormpath module](https://github.com/stormpath/stormpath-sdk-node) that contains the changes from PR https://github.com/stormpath/stormpath-sdk-node/pull/213.

#### What this fixes

* Integration config now in integration instead of SDK.
* Social providers now enriched automatically instead of relying on hard-coded defaults in config. Because of this we now have to check if the provider exists before using it. See: https://github.com/stormpath/express-stormpath/compare/refactor-client-config?expand=1#diff-ef4e9b51bbe32bc78fc66cfd8d410182R127.

(additional)

* SDK errors are now emitted through express app channel `stormpath.error`.

#### How to verify

1. Run `$ npm test`.
2. Either create a new, or use an existing `express-stormpath` app.
3. Assert that both configuration options `website` and `api` enable the correct endpoints for those options.
4. Assert that social providers are being populated by social provider directories mapped to your application.

### CODE REVIEW

**Comment by robertjd** [comment]
I was reviewing this PR, and I'm happy with what I'm seeing.  I updated the Node SDK dependency and merged in master.  The get-user test is going to need to be fixed, it is depending on the old way of loading a default configuration

**Comment by robertjd** [comment]
There is one dilemma I’d like to point out.  With this PR, we don’t bind express routes until after the configuration is resolved, which is async.  This is annoying if you’re using a live-reload environment, because a route will return 404 before the configuration is loaded, but then return our framework response once the configuration is loaded.

We should decide if we care about this use-case.  But I’ll stay that as developer who supports our Angular SDK (and the related example application) I’m frustrated by this break in my live-reload workflow.

If we were to support this use case, we’d have to synchronously load the basic configuration that tells us which routes we want to bind to.  We would then eagerly bind to the routes, then assume that any calls to the underlying SDK will be deferred until the client is actually ready (the config has been resolved).  This appears to be the case with the 0.13.0 release of the Node SDK.

**Comment by typerandom** [comment]
@robertjd Could you explain what you mean with that the "get-user test needs to be fixed"?

Also, when it comes to live-reload, what we could do is do so that the middleware blocks until the SDK is ready. I.e. we inject the middleware, but does not continue until the middleware is ready.

I'll start working on a fix for this and let you know when you can review it :)

**Comment by robertjd** [comment]
> @robertjd Could you explain what you mean with that the "get-user test needs to be fixed"?

These tests are failing right now, because they're trying to require the old config.json:

https://github.com/stormpath/express-stormpath/blob/master/test/helpers/test-get-user.js

I love fixing tests, so I will fix those.

> Also, when it comes to live-reload, what we could do is do so that the middleware waits until the SDK is ready. I.e. we inject the middleware, but doesn't let the middleware continue until the SDK is ready.

That is what I was thinking.  Do you have an idea of how you'll go about it?  It seems that we'd have to do some synchronous loading of any provided configuration files.

**Comment by typerandom** [comment]
>These tests are failing right now, because they're trying to require the old config.json. I love fixing tests, so I will fix those.

I hate fixing tests:) But awesome, I'll do my things and if you get time to fix them then please do. Else I'll just do it:)

>That is what I was thinking. Do you have an idea of how you'll go about it? It seems that we'd have to do some synchronous loading of any provided configuration files.

We can still do everything async. All we need to do is that if the SDK isn't ready, then await the `ready` event before calling next on our middleware. I.e. so that all requests that use our middleware will queue until we get our `ready` event.

**Comment by robertjd** [comment]
I want to review this but won't have time tonight, will pickup tomorrow.

**Comment by typerandom** [comment]
I have now updated so that the new Stormpath module v0.13.2 is in use. Just noticed that I had a typo in my commit message saying version 0.13.1, but it should be 0.13.2 :)

**Comment by robertjd** [comment]
I like the YAML file, that is a lot easier to read than JSON.. :)

There is a failing case for the unerlying option parser, can you take a look at this case I wrote?

https://github.com/stormpath/stormpath-node-config/commit/a955f6bd9670c8a0c338ca6df7875558ec9f0862

Thanks!

**Comment by typerandom** [comment]
Yeah, I also like it a lot more :) Ah, this is related to this issue: https://github.com/stormpath/express-stormpath/issues/106

Right?

**Comment by robertjd** [comment]
Yes!  I was trying to find that.. but my mind knew it existed, hence the test :)

**Comment by typerandom** [comment]
Awesome! I have an idea on how to fix that, so will look into it :)

**Comment by robertjd** [comment]
I merged https://github.com/stormpath/stormpath-node-config/pull/3 - do we need to do much here?

File: package.json [hunk 1]
```diff
@@ -28,13 +28,13 @@
     "csurf": "^1.8.0",
     "deep-extend": "^0.4.0",
     "express": "^4.10.6",
-    "express-stormpath-config": "0.0.2",
```
Comment (robertjd) [inline package.json]: Do we delete this package after this PR is merged?

File: test/controllers/test-email-verification.js [hunk 2]
```diff
@@ -38,20 +38,32 @@ describe('email verification', function() {
...
+        stormpathApplication = app;
+
+        helpers.setEmailVerificationStatus(stormpathApplication, 'ENABLED', done);
+      });
     });
   });
 
   after(function(done) {
-    helpers.destroyApplication(stormpathApplication, done);
+    if (stormpathApplication) {
```
Comment (robertjd) [inline test/controllers/test-email-verification.js]: Why this if clause?

**Comment by robertjd** [comment]
@typerandom please merge in master to handle conflicts, then ping me for one last review before we merge this

File: package.json (hunk 1 above)
Comment (typerandom) [inline package.json]: Yes, we can do that.

File: test/controllers/test-email-verification.js (hunk 2 above)
Comment (typerandom) [inline test/controllers/test-email-verification.js]: I can't remember exactly. I'm just going to remove it:)

**Comment by typerandom** [comment]
@robertjd It's merged in:) Please review.

**Comment by robertjd** [comment]
So close!  I'm signing off on the code quality (disclaimer: I'm ignoring all the declared but un-used variables, but only because we haven't yet merged our eslint config :) )

But now I've got an error, @typerandom can you help me reproduce?   I'm getting an error from the extend library, using my demo angular application.

Please clone this repo and enter this folder and follow the instructions to setup this demo application:

https://github.com/stormpath/stormpath-sdk-angularjs/tree/master/example/dashboard-app

Then use `npm link express-stormpath` to link that demo to this branch.

When I run `grunt serve`, I'm getting this error:

```
/Users/robert/projects/express-stormpath/node_modules/stormpath-config/node_modules/cloneextend/index.js:48
 b_pos=context.indexOf(b);
               ^
TypeError: undefined is not a function
    at extend (/Users/robert/projects/express-stormpath/node_modules/stormpath-config/node_modules/cloneextend/index.js:48:16)
    at /Users/robert/projects/express-stormpath/node_modules/stormpath-config/lib/strategy/EnrichIntegrationFromRemoteConfigStrategy.js:83:15
    at Object.wrapGetResourceResponse [as _wrapGetResourceResponse] (/Users/robert/projects/express-stormpath/node_modules/stormpath/lib/ds/DataStore.js:89:12)
    at onGetResourceRequestResult (/Users/robert/projects/express-stormpath/node_modules/stormpath/lib/ds/DataStore.js:278:13)
    at Request.onRequestResult [as _callback] (/Users/robert/projects/express-stormpath/node_modules/stormpath/lib/ds/RequestExecutor.js:110:7)
    at Request.self.callback (/Users/robert/projects/express-stormpath/node_modules/stormpath/node_modules/request/request.js:123:22)
    at Request.emit (events.js:110:17)
    at Request.<anonymous> (/Users/robert/projects/express-stormpath/node_modules/stormpath/node_modules/request/request.js:1047:14)
    at Request.emit (events.js:129:20)
    at IncomingMessage.<anonymous> (/Users/robert/projects/express-stormpath/node_modules/stormpath/node_modules/request/request.js:998:12)
```

**Comment by typerandom** [comment]
@robertjd I'm looking into this right now.

**Comment by typerandom** [comment]
@robertjd This issue is now fixed:)

Btw, the un-used variables, I guess you referring to things in tests? Because everywhere else we should use strict mode and and I haven't seen those warnings. Or how are you seeing these, did you run eslint over or are you using something else?

**Comment by robertjd** [comment]
Awesome, thanks!  We can merge this PR after we fix this issue:

https://github.com/stormpath/stormpath-sdk-node/pull/253

Regarding un-used vars, there are some in lib/stormpath.js.  I've taken the .eslintrc from the eslint branch and put it in my home folder, so it's being used by default now in my sublime :)

**Comment by robertjd** [comment]
@typerandom woot!

**Pull request merged**